src/dmd/must_use.d

```diff
@@ -0,0 +1,203 @@
+{
+    import dmd.id : Id;
+
+    if (sym.userAttribDecl is null || sym.userAttribDecl.atts is null)
+        return;
+
+    // Can't use foreachUda (and by extension hasMustUseAttribute) while
+    // semantic analysis of `sym` is still in progress
+    // TODO: factor out common code from this function and checkGNUABITag
+    foreach (exp; *sym.userAttribDecl.atts)
+    {
+        if (isMustUseAttribute(exp))
+        {
+            if (sym.isFuncDeclaration())
+            {
+                sym.error("`@%s` on functions is reserved for future use",
+                    Id.udaMustUse.toChars());
+                sym.errors = true;
+            }
+            else if (sym.isClassDeclaration() || sym.isInterfaceDeclaration() || sym.isEnumDeclaration())
+            {
+                sym.error("`@%s` on `%s` types is reserved for future use",
+                    Id.udaMustUse.toChars(), sym.kind());
+                sym.errors = true;
+            }
+        }
+    }
+}
+
+/**
+ * Returns true if the given expression is an assignment, either simple (a = b)
+ * or compound (a += b, etc).
+ */
+private bool isAssignment(Expression e)
+{
+    if (e.isAssignExp || e.isBinAssignExp)
+        return true;
+    if (auto ce = e.isCallExp())
+    {
+        auto fd = ce.f;
+        auto id = fd ? fd.ident : null;
+        // opXXX are reserved identifiers, so it's ok to match too much here
+        if (id && id.toString().startsWith("op") && id.toString().endsWith("Assign"))
+            return true;
+    }
+    return false;
+}
+
+/**
+ * Returns true if `s` starts with `prefix`, false otherwise.
+ */
+private bool startsWith(const(char)[] s, const(char)[] prefix)
+{
+    if (s.length < prefix.length)
+        return false;
+    return s[0 .. prefix.length] == prefix;
+}
+
+/**
+ * Returns true if `s` ends with `suffix`, false otherwise.
+ */
+private bool endsWith(const(char)[] s, const(char)[] suffix)
+{
+    if (s.length < suffix.length)
+        return false;
+    return s[$ - suffix.length .. $] == suffix;
+}
+
+/**
+ * Returns true if the given expression is an increment (++) or decrement (--).
+ */
+private bool isIncrementOrDecrement(Expression e)
+{
+    import dmd.dtemplate : isExpression;
+    import dmd.globals : Loc;
+    import dmd.tokens : EXP;
+
+    if (e.op == EXP.plusPlus
+        || e.op == EXP.minusMinus
+        || e.op == EXP.prePlusPlus
+        || e.op == EXP.preMinusMinus)
+        return true;
+    if (auto call = e.isCallExp())
+    {
+        // Check for overloaded preincrement
+        // e.g., a.opUnary!"++"
+        auto fd = call.f;
+        auto id = fd ? fd.ident : null;
+        if (id && id.toString() == "opUnary")
+        {
+            auto ti = fd.parent ? fd.parent.isTemplateInstance() : null;
+            auto tiargs = ti ? ti.tiargs : null;
+            if (tiargs && tiargs.length >= 1 && (*tiargs)[0].isExpression())
+            {
+                auto op = (cast(Expression) (*tiargs)[0]).isStringExp();
```

Comment:
```suggestion
                auto op = (*tiargs)[0]).isStringExp();
```

Reply:
Can't call `isStringExp` on a `RootObject`.

Reply:
check EXP explicitly? Some of these have already been implemented as UFCS helpers lying around the code

Reply:
Again, `RootObject` doesn't have an `EXP`. You have to `dyncast` it or use a helper function like `toExpression()` before you can do anything else.

The best I can do here is refactor the code so the result of `(*tiargs)[0].isExpression()` is saved in a variable, which would eliminate the need for the naked `cast`.

Reply:
Please disregard my initial comment, I had misread the braces.